### ISSUE TITLE
Attention module not fused for custom GPT2 model

### ISSUE THREAD

**Issue by dave-rtzr** [microsoft/onnxruntime#10339]
Hi,

This issue is not about bug.

I would like to fuse the graph of my model, so that I can get a speed up as illustrated [here](https://github.com/microsoft/onnxruntime/tree/master/onnxruntime/python/tools/transformers#model-optimizer)
One thing that differs from the example is that I am not using GPT2LMHeadModel, which I confirmed that the fusion works as expected,
but I am trying to fuse a custom GPT2 Model as follows
This model is used for estimating the likelihood of a sentence.

```python
class MyGPT2Model(GPT2LMHeadModel):
  def __init__(self, config):
    super().__init__(config)
    self.pad_token_id = 3 # for example
    
  def forward(self, input_ids):
    attention_mask = torch.ones_like(input_ids)
    attention_mask[ input_ids == self.pad_token_id ] = 0
    logits = super().forward(input_ids=input_ids, attention_mask=attention_mask, use_cache=False, return_dict=False)[0] #logits
    logits = logits[ : , :-1 , : ]
    labels = input_ids[ : , 1: ]

    return MyGPT2Model.postprocess(logits, labels, self.pad_token_id)
    
  @staticmethod  
  def postprocess(logits, labels, pad_token_id: int):
    log_probs = torch.nn.functional.log_softmax(logits, dim=-1)
    out = log_probs.gather(dim=2, index=labels.unsqueeze(2)).squeeze(2)
    out[ labels == pad_token_id ] = 0
    return out.sum(dim=1)
```

I designed my model as above since gpt2_helper.py had a similar model, which I assumed the optimization would succeed.
I exported above model as follows.

```python
model = MyGPT2Model.from_pretrained("gpt2").to("cuda:0")
model.eval()
bsz, seq_len = 256, 64
x = torch.randint(0, 1000, (bsz, seq_len), dtype=torch.long, device="cuda:0") # 1000 is less than the vocab size

with torch.inference_mode():
  torch.onnx.export(
                model,
                args=(x,),
                f="/workspace/model.onnx",
                input_names=["input__0"],
                output_names=["output__0"],
                opset_version=11,
                dynamic_axes={
                    "input__0": {0: "batch_size", 1: "seq_len"},
                    "output__0": {0: "batch_size"},
                },
                do_constant_folding=True,
            )
```

after exporting model to onnx using torch.onnx.export with opset 11, and try to optimize the model using the following command,

```python
root@ef62a7f0723c:/workspace# python -m onnxruntime.transformers.optimizer --input ./model.onnx --output ./out.onnx --model_type gpt2 --num_heads 12 --hidden_size 768 --input_int32 --float16 --use_gpu --opt_level 0
```
I constantly get the following error

```bash
remove_useless_cast_nodes: Removed 15 Cast nodes with output type same as input
               apply: Fused LayerNormalization count: 25
               apply: Fused FastGelu count: 12
remove_useless_reshape_nodes: Remove reshape node Reshape_17 since its input shape is same as output: ['batch_size', 'seq_len']
remove_useless_reshape_nodes: Remove reshape node Reshape_26 since its input shape is same as output: [1, 'seq_len']
remove_useless_reshape_nodes: Remove reshape node Reshape_29 since its input shape is same as output: ['batch_size', 'seq_len']
remove_useless_reshape_nodes: Remove reshape node Reshape_2365 since its input shape is same as output: ['batch_size', 'seq_len', 768]
         prune_graph: Graph pruned: 0 inputs, 0 outputs and 742 nodes are removed
         postprocess: postprocess: remove Reshape count:72
               apply: Fused FastGelu(add bias) count: 12
            optimize: opset verion: 11
change_graph_inputs_to_int32: Graph inputs are changed to int32. Added 1 Cast nodes, and removed 0 Cast nodes.
  save_model_to_file: Sort graphs in topological order
  save_model_to_file: Model saved to ./out.onnx
get_fused_operator_statistics: Optimized operators:{'EmbedLayerNormalization': 0, 'Attention': 0, 'Gelu': 0, 'FastGelu': 12, 'BiasGelu': 0, 'LayerNormalization': 25, 'SkipLayerNormalization': 0}
  is_fully_optimized: Attention not fused
                main: The model has been optimized.
```

I am working on nvcr.io/nvidia/pytorch:21.12-py3, with onnxruntime-gpu==1.10.0, transformers==4.13.0 manually installed.

It seems like the problem is not caused by the postprocessing function, since just returning the logits also yields the same error.
I would appreciate any ideas on this problem

**Comment by tianleiwu** [comment]
It is likely that the graph is slightly different (caused by different version of PyTorch or transformers, or your modifications), that caused Attention not fused. Please refer to our developer guide:
https://github.com/microsoft/onnxruntime/blob/master/onnxruntime/python/tools/transformers/Dev_Guide.md

First, you can try pass attention_mask as input to see whether it could resolve the issue.

@dave-rtzr, you can debug a little into the following fusion to see which part of subgraph is not matched and modify it accordingly: https://github.com/microsoft/onnxruntime/blob/master/onnxruntime/python/tools/transformers/fusion_gpt_attention_no_past.py

Feel free to contribute some improvements to make the fusion more robust.
